### PR TITLE
Note: Implement Storage & Store as JSON

### LINKED ISSUE #92: As a user I can save the information I enter

### LINKED ISSUE #177: As a user, I can transfer the entered data across different computers

### LINKED ISSUE #178: Store Notes in JSON

### PR DESCRIPTION
**Pull request by rtshkmr**
## Description

Implements Notes storage. Follows closely with storage skeleton as per #198 and #185 

Fixes #178 

## Testing

- Run any client related notes and tagging and it will persist on subsequent runs of TBM. 

## Progress: 
- **Done: update `Addressbook#getNoteList()`** : currently chose to fetch clientNotes by iterating thru all the uniqueClients. ref to [this comment](https://github.com/AY2021S1-CS2103T-F11-4/tp/pull/235#discussion_r511558049) on why.
-  **Done: update `JsonAdaptedClientNote`:** currently stores a set of `JsonAdaptedTag` inside `JsonAdaptedNote`. This is because the `toModelType()` function needs to be able to recreate clientNote objects from its Json form, which requires creating the correct `Tag` objects associated to those clientNotes.
- **Serialization and Deserialisation is done properly nnoww**
image

* related todo: haveToModify addressbook tests, they're failing now (haven't pushed this commit yet)
image

* take note: the collection of unique tags doesn't seem to get updated when tagged client note is added, should address this in a separate issue actually

### Questions: 
1. how come this constructor is never used? 
image

WIP

### CODE REVIEW

File: src/main/java/seedu/address/model/AddressBook.java [hunk 1]
```diff
@@ -76,6 +77,14 @@ public void setNotes(List<Note> notes) {
...
             if (note.isClientNote()) {
                 // handle client notes
+                // todo:  =======================================================
+                //        decided to not store the client notes in a separete set
+                //        because the notes are going to be stored within clients
+                //        itself. The setClient function aldy does this. Adding in
+                //        a collection of client notes in this class would mean that
+                //        collection needs to be constantly updated. As such, it
+                //        might be good enough to just modify the getNoteList method
+                //        ===========================================================
```
Comment (rtshkmr) [inline src/main/java/seedu/address/model/AddressBook.java:87]: @raysonkoh I'm sticking to this until it comes off as a problem. 

(for my own reference, relevant alernative: [this commit for a deleted PR](https://github.com/rtshkmr/tp/commit/e30108c552aad0b342cb65e5bc39af07308b50c3) )

File: src/main/java/seedu/address/storage/JsonSerializableAddressBook.java [hunk 2]
```diff
@@ -31,7 +31,7 @@
      */
     @JsonCreator
     public JsonSerializableAddressBook(@JsonProperty("clients") List<JsonAdaptedClient> clients,
-            @JsonProperty("notes") List<JsonAdaptedNote> notes) {
+            @JsonProperty("allNotes") List<JsonAdaptedNote> notes) {
```
Comment (raysonkoh) [inline src/main/java/seedu/address/storage/JsonSerializableAddressBook.java]: Found the bug. Should be `"notes"`

**Review by raysonkoh — DISMISSED** [review]

File: src/main/java/seedu/address/storage/JsonSerializableAddressBook.java (hunk 2 above)
Comment (rtshkmr) [reply src/main/java/seedu/address/storage/JsonSerializableAddressBook.java]: ❤️ thanks m8

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -177,6 +178,7 @@ public void addClientNote(Client target, Note clientNote) {
         requireAllNonNull(target, clientNote);
         target.addClientNote(clientNote);
         // update the TagNoteMap:
+        // todo: update address book to reflect the new note added?
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/ModelManager.java]: probably no need for this, since we're mutating the client directly

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 4]
```diff
@@ -76,6 +83,8 @@ public JsonAdaptedClient(Client source) {
         contractExpiryDate = source.getContractExpiryDate().value;
         lastModifiedInstant = source.getLastModifiedInstant().toString();
         tagged.addAll(source.getTags().stream().map(JsonAdaptedTag::new).collect(Collectors.toList()));
+        clientNotes
+            .addAll(source.getClientNotes().stream().map(JsonAdaptedNote::new).collect(Collectors.toSet()));
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java]: spacing here should be 8 spaces i think

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 5]
```diff
@@ -88,72 +97,81 @@ public Client toModelType() throws IllegalValueException {
...
         }
         if (!Name.isValidName(name)) {
             throw new IllegalValueException(Name.MESSAGE_CONSTRAINTS);
         }
         final Name modelName = new Name(name);
 
         if (phone == null) {
-            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Phone.class.getSimpleName()));
+            throw new IllegalValueException(
+                String.format(MISSING_FIELD_MESSAGE_FORMAT, Phone.class.getSimpleName()));
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java]: Not too sure why these were reformatted, should probably set ur auto-formatter to reformat only when chars > 120 in a line, same for everything below. Also, I think the second line should be 8 spaces instead of 4, when doing a linebreak

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 6]
```diff
@@ -162,8 +180,13 @@ public Client toModelType() throws IllegalValueException {
...
         final Set<Tag> modelTags = new HashSet<>(clientTags);
-        return new Client(modelName, modelPhone, modelEmail, modelAddress, modelCountry,
-            modelTimezone, modelContractExpiryContractExpiryDate, modelLastModifiedInstant, modelTags);
+        Client freshClient =
+                new Client(modelName, modelPhone, modelEmail, modelAddress, modelCountry, modelTimezone,
+                    modelContractExpiryContractExpiryDate, modelLastModifiedInstant, modelTags);
+        for (Note note : clientNotes) {
+            freshClient.addClientNote(note);
+        }
+        return freshClient;
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java]: probably should name this modelClient instead of freshClient

File: src/main/java/seedu/address/storage/JsonAdaptedNote.java [hunk 7]
```diff
@@ -50,9 +62,15 @@ public boolean isClientNote() {
      *
      * @return The correct Note object being represented by this json note.
      */
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedNote.java]: Should probably change this javadoc to match those of other `toModelType()` functions like JsonAdaptedClient etc.

File: src/test/java/seedu/address/storage/JsonAdaptedClientTest.java [hunk 8]
```diff
@@ -2,25 +2,32 @@
```
Comment (qwoprocks) [inline src/test/java/seedu/address/storage/JsonAdaptedClientTest.java:2]: Not sure why the formatting in this file changed so much also, would suggest to change ur autoformatter settings to match the project's code style, or not use the autoformatter, if not unrelated lines will keep changing in your commits

File: src/test/java/seedu/address/storage/JsonAdaptedNoteTest.java [hunk 9]
```diff
@@ -3,23 +3,37 @@
...
-    public void toModelType_validCountryNote_returnsCountryNote() {
-        JsonAdaptedNote jsonAdaptedNote = new JsonAdaptedNote(new CountryNote("some note", new Country("SG")));
+    @Test public void toModelType_validCountryNote_returnsCountryNote() throws IllegalValueException {
+        JsonAdaptedNote jsonAdaptedNote =
+                new JsonAdaptedNote(new CountryNote("some note", new Country("SG")));
         assertFalse(jsonAdaptedNote.toModelType().isClientNote());
     }
+    /* todo: future tests:
+    *       1. how to test the @JsonCreator annotated constructor? in the first place it's not used.
+    * */
```
Comment (qwoprocks) [inline src/test/java/seedu/address/storage/JsonAdaptedNoteTest.java:37]: formatting issues with this file, and for the testing of the JsonCreator, you can refer to JsonAdaptedClientTest.java, also it only appears to be not used because a third party library (jackson), and not our own code, is calling it, just like how the FXML notation variables don't seemed to be used

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json [hunk 10]
```diff
@@ -64,5 +82,23 @@
...
+    "contents" : "his son is the devil's incarnate",
+    "countryCode" : "NULL_CC",
+    "tags" : [ "turnOff" ]
+  },
+    {
+    "contents" : "snores a lot",
+    "countryCode" : "NULL_CC",
+    "tags" : [ "turnOff" ]
+  }
+  ]
```
Comment (qwoprocks) [inline src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: these should be on the same line

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json [hunk 11]
```diff
@@ -64,5 +82,23 @@
...
+  }, {
+    "contents" : "secret to the best krabby patty is in the J.U.I.C.E.",
+    "countryCode" : "NULL_CC",
+    "tags" : [ "seduction" ]
+  }, {
+    "contents" : "his son is the devil's incarnate",
+    "countryCode" : "NULL_CC",
+    "tags" : [ "turnOff" ]
+  },
+    {
```
Comment (qwoprocks) [inline src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: these should be on the same line

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json [hunk 12]
```diff
@@ -17,7 +26,16 @@
...
+    "tagged" : [ "owesMoney", "friends" ],
+    "clientNotes" : [ {
+      "contents" : "his son is the devil's incarnate",
+      "countryCode" : "NULL_CC",
+      "tags" : [ "turnOff" ]
+    }, {
+      "contents" : "snores a lot",
+      "countryCode" : "NULL_CC",
+      "tags" : [ "turnOff" ]
+    } ]
```
Comment (qwoprocks) [inline src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: no space between `[` and `{`

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json [hunk 13]
```diff
@@ -8,7 +8,16 @@
     "country" : "SG",
     "timezone": "GMT+8",
     "contractExpiryDate": "1-4-2021",
-    "tagged" : [ "friends" ]
+    "tagged" : [ "friends" ],
+    "clientNotes" : [ {
+      "contents" : "lives in a pineapple",
+      "countryCode" : "NULL_CC",
```
Comment (qwoprocks) [inline src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: actually for client notes, is it possible to not store the countryCode in the first place? Seems a bit wasteful and unneeded. JsonCreator and toModelType() can probably handle the lack of the countryCode programatically. @raysonkoh any comments about this?

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json (hunk 13 above)
Comment (raysonkoh) [reply src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: Will create an issue for this. Currently it is working and it is quite close to end of v1.3, changing this might introduce unintentional bugs.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 3 above)
Comment (rtshkmr) [reply src/main/java/seedu/address/model/ModelManager.java]: okay, understood 👍

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java (hunk 5 above)
Comment (rtshkmr) [reply src/main/java/seedu/address/storage/JsonAdaptedClient.java]: Yeap autoformatter :') Changed it.

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json (hunk 13 above)
Comment (qwoprocks) [reply src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: @raysonkoh I think if we change it, we should change before the end of v1.3, since we have our bug-finding practice coming up and people can help catch the bugs if any. If we change after v1.3, we might introduce some bugs that no one finds until the final bug-finding test.

File: src/test/java/seedu/address/storage/JsonAdaptedNoteTest.java (hunk 9 above)
Comment (rtshkmr) [reply src/test/java/seedu/address/storage/JsonAdaptedNoteTest.java:37]: psps, autoformatter messed up my formatting, I havenow correctly changed it

btw the `JsonAdaptedClientTest.java` doesn't have tests for the JsonCreater right...

image

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json (hunk 12 above)
Comment (rtshkmr) [reply src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: does this look better? tried the autoformatting and it makes things into a pretty tree-like format

image

File: src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json [hunk 14]
```diff
@@ -1,68 +1,116 @@
...
-    "phone" : "9482442",
-    "email" : "anna@example.com",
-    "address" : "4th street",
-    "country" : "US",
-    "timezone": "GMT-4",
-    "contractExpiryDate": "2-8-2021",
-    "tagged" : [ ]
-  } ],
-  "notes": []
+  "clients": [
```
Comment (qwoprocks) [inline src/test/data/JsonSerializableAddressBookTest/typicalClientsAddressBook.json]: the formatting is still abit off, it should be `[ {` instead of `[<br>{` and `}, {` instead of `},<br>{`

File: src/test/java/seedu/address/storage/JsonAdaptedClientTest.java (hunk 8 above)
Comment (qwoprocks) [reply src/test/java/seedu/address/storage/JsonAdaptedClientTest.java:2]: ^

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 15]
```diff
@@ -32,24 +33,29 @@
 
     private final String name;
     private final String phone;
+
+
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java]: extra new lines, also some formatting issues with this file, where lines <120 chars are being broken up for no reason, leading to changes in random lines.

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 16]
```diff
@@ -76,6 +85,8 @@ public JsonAdaptedClient(Client source) {
         contractExpiryDate = source.getContractExpiryDate().value;
         lastModifiedInstant = source.getLastModifiedInstant().toString();
         tagged.addAll(source.getTags().stream().map(JsonAdaptedTag::new).collect(Collectors.toList()));
+        clientNotes
+                .addAll(source.getClientNotes().stream().map(JsonAdaptedNote::new).collect(Collectors.toSet()));
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java]: These 2 lines are <120 chars, they can be on the same line. Also, breaking it up at the first . makes it inconsistent with the rest of the code base, and harder to read.

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 17]
```diff
@@ -162,8 +182,13 @@ public Client toModelType() throws IllegalValueException {
         }
 
         final Set<Tag> modelTags = new HashSet<>(clientTags);
-        return new Client(modelName, modelPhone, modelEmail, modelAddress, modelCountry,
-            modelTimezone, modelContractExpiryContractExpiryDate, modelLastModifiedInstant, modelTags);
+        Client modelClient =
+                new Client(modelName, modelPhone, modelEmail, modelAddress, modelCountry, modelTimezone,
+                    modelContractExpiryContractExpiryDate, modelLastModifiedInstant, modelTags);
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java]: formatting, breaking a line up after `=` makes it inconsistent with the rest of the code base

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: src/test/java/seedu/address/storage/JsonAdaptedClientTest.java (hunk 8 above)
Comment (rtshkmr) [reply src/test/java/seedu/address/storage/JsonAdaptedClientTest.java:2]: updated

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java (hunk 15 above)
Comment (rtshkmr) [reply src/main/java/seedu/address/storage/JsonAdaptedClient.java]: resolved

**Review by qwoprocks — APPROVED** [review]
LGTM
